pennylane_lightning/lightning_gpu/_adjoint_jacobian.py

```diff
@@ -0,0 +1,247 @@
+
+
+import numpy as np
+from pennylane import BasisState, StatePrep
+from pennylane.operation import Operation
+from pennylane.tape import QuantumTape
+from scipy.sparse import csr_matrix
+
+# pylint: disable=ungrouped-imports
+from pennylane_lightning.core._adjoint_jacobian_base import LightningBaseAdjointJacobian
+from pennylane_lightning.core._serialize import QuantumScriptSerializer
+
+
+class LightningGPUAdjointJacobian(LightningBaseAdjointJacobian):
+    """Check and execute the adjoint Jacobian differentiation method.
+
+    Args:
+        qubit_state(LightningGPUStateVector): State Vector to calculate the adjoint Jacobian with.
+        batch_obs(bool): If serialized tape is to be batched or not. For Lightning GPU,
+            if `batch_obs=False` the computation requires more memory and is faster,
+            while `batch_obs=True` allows a larger number of qubits simulation
+            at the expense of high computational cost. Defaults to False
+    """
+
+    # pylint: disable=too-few-public-methods
+
+    def __init__(
+        self,
+        qubit_state: LightningGPUStateVector,  # pylint: disable=undefined-variable
+        batch_obs: bool = False,
+    ) -> None:
+
+        super().__init__(qubit_state, batch_obs)
+
+        self._dp = DevPool()
+
+        self._use_mpi = qubit_state._mpi_handler.use_mpi
+
+        if self._use_mpi:
+            self._mpi_handler = qubit_state._mpi_handler
+
+        # Initialize the C++ binds
+        self._jacobian_lightning, self._create_ops_list_lightning = self._adjoint_jacobian_dtype()
+
+        # Warning about performance with MPI and batch observation
+        if self._use_mpi and not self._batch_obs:
+            warn(
+                "Using LightningGPU with `batch_obs=False` and `use_mpi=True` has the limitation of requiring more memory. If you want to allocate larger number of qubits use the option `batch_obs=True`"
+                "For more information Check out the section `Parallel adjoint differentiation support` in our website https://docs.pennylane.ai/projects/lightning/en/stable/lightning_gpu/device.html for more details.",
+                RuntimeWarning,
+            )
+
+    def _adjoint_jacobian_dtype(self):
+        """Binding to Lightning GPU Adjoint Jacobian C++ class.
+
+        Returns: the AdjointJacobian class
+        """
+        if self._use_mpi:
+            if not MPI_SUPPORT:
+                warn(str(mpi_error), UserWarning)
+
+            jacobian_lightning = (
+                AdjointJacobianMPIC64() if self.dtype == np.complex64 else AdjointJacobianMPIC128()
+            )
+            create_ops_list_lightning = (
+                create_ops_listMPIC64 if self.dtype == np.complex64 else create_ops_listMPIC128
+            )
+            return jacobian_lightning, create_ops_list_lightning
+
+        # without MPI
+        jacobian_lightning = (
+            AdjointJacobianC64() if self.dtype == np.complex64 else AdjointJacobianC128()
```

Comment:
Can we move also these to different methods?

Reply:
I think it will be more confusing if we add a second method for defining the bind.